File: src/grid/resource-manager.ts
import { NodeType } from '../models/node-type';
import { sleep } from '../utils/sleep';
import { Job } from './job';
import { Message, MessageType } from './message';
import { GridNode, NodeStatus } from './node';
import { Socket } from './socket';

export class ResourceManagerNode extends GridNode {
    private static readonly SCHEDULER_LOST = 'A scheduler should be stored for this job';

    /**
     * Collection of workers that are registered to this resource manager.
     */
    private workers = new Set<Socket>();

    /**
     * Collection of jobs that have to be processed by this resource manager.
     */
    private jobs = new Set<Job>();

    /**
     * Map used to keep track of which scheduler to return a certain job to.
     */
    private schedulerMap = new Map<Job, Socket>();

    /**
     * Map used to keep track of messages that are unconfirmed.
     */
    // private confirmed = new Map<Job, boolean>();

    constructor() {
        super(NodeType.ResourceManager);
    }

    /**
     * Registers the provided worker as usable by this resource manager.
     * @param worker The worker socket to register.
     */
    public registerWorker(worker: Socket): this {
        this.workers.add(worker);
        return this;
    }

    /**
     * Registers the provided workers as usable by this resource manager.
     * @param workers The worker sockets to register.
     */
    public registerWorkers(workers: Socket[]): this {
        workers.forEach(worker => this.registerWorker(worker));
        return this;
    }

    /**
     * The run method is constantly triggered from the base class.
     */
    public async run(): Promise<void> {
        // For every active-unfinished job periodically check if the worker is
        // still alive
        // for (let i = 0; i < this.jobs.size; i++) {
        //     if (this.jobs[i].status !== JobStatus.Running) continue;
        //     const worker = this.jobs[i].workerSocket;
        //     this.pingWorker(worker);
        // }
        await sleep(200);
    }

    /**
     * Executed when a message is received.
     * @param message The message.
     */
    public onMessage(message: Message): void {
        // Ignore every message if dead
        if (this.status === NodeStatus.Dead) return;

        switch (message.from.type) {
            case NodeType.Scheduler:
                this.onSchedulerMessage(message);
                break;
            case NodeType.Worker:
                this.onWorkerMessage(message);
                break;
        }
    }

    /**
     * Executed when a message is received from a scheduler.
     * @param message The message.
     */
    private onSchedulerMessage(message: Message): void {
        switch (message.type) {
            case MessageType.Request:
                this.onSchedulerRequest(message);
                break;
            case MessageType.Acknowledgement:
                this.onSchedulerAcknowledgement(message);
                break;
            // case MessageType.Ping:
            //     this.onSchedulerPing(message);
            //     break;
        }
    }

    /**
     * Executed when a message is received from a worker.
     * @param message The message.
     */
    private onWorkerMessage(message: Message): void {
        switch (message.type) {
            case MessageType.Confirmation:
                this.onWorkerConfirmation(message);
                break;
            case MessageType.Result:
                this.onWorkerResult(message);
                break;
            // case MessageType.Ping:
            //     this.onWorkerPing(message);
            //     break;
        }
    }

    /**
     * Executed when a request message is received from a scheduler.
     * @param message The message.
     */
    private onSchedulerRequest(message: Message): void {
        const { job } = message;

        // Update job count
        this.jobs.add(job);
        this.setJobCount(this.jobs.size);

        // Keep job -> scheduler mapping
        this.schedulerMap.set(job, message.from);

        // Send confirmation to the scheduler (fig 2, step 6)
        const confirmation = new Message(this, MessageType.Confirmation, job);
        message.from.send(confirmation);

        // Send job to worker
        const request = new Message(this, MessageType.Request, job);
        this.sendToWorker(request);
    }

    /**
     * Executed when an acknowledgement message is received from a scheduler.
     * @param message The message.
     */
    private onSchedulerAcknowledgement(message: Message): void {
        // Remove from job -> scheduler mapping
        this.schedulerMap.delete(message.job);
    }

    /**
     * Executed when a ping message is received from a scheduler.
     * @param message The message.
     */
    private onSchedulerPing(message: Message): void {}

    /**
     * Executed when a confirmation message is received from a worker.
     * @param message The message.
     */
    private onWorkerConfirmation(message: Message): void {
        // Register as confirmed
        // this.confirmed.set(message.job, true);
    }

    /**
     * Executed when a result message is received from a worker. A confirmation
     * message is sent to the worker and the job result is returned to the
     * scheduler.
     * @param message The message.
     */
    private onWorkerResult(message: Message): void {
        const { job } = message;
        this.workers.add(message.from);

        // Confirm the result for a worker
        // const confirmation = new Message(this, MessageType.Confirmation, job);
        // message.from.send(confirmation);

        // Send job result to the scheduler (fig 2, step 10)
        const scheduler = this.schedulerMap.get(job);
        if (!scheduler) throw new Error(ResourceManagerNode.SCHEDULER_LOST);
        const result = new Message(this, MessageType.Result, job);
        scheduler.send(result);

        // Update job count
        this.jobs.delete(job);
        this.setJobCount(this.jobs.size);
    }

    /**
     * Picks an available worker and sends the message to it. If no confirmation
     * is received within 5 seconds the message is sent to another worker.
     * @param message The message to send
     */
    private sendToWorker(message: Message): void {
        const { job } = message;

        // Pick a worker & send the message
        const worker = [...this.workers][0];

        // If there is no available worker it will retry after 5 seconds.
        if (worker) {
            worker.send(message);

            // Remove from available workers & register job as unconfirmed
            this.workers.delete(worker);
            // this.confirmed.set(job, false);
        }

        // Retry using a different worker if no confirmation is received
        // within 5 seconds
        // setTimeout(() => {
        //     if (!this.confirmed.get(job)) {
        //         console.log('test');
        //         this.sendToWorker(message);
        //     } else this.confirmed.delete(job);
        // }, 5000);
    }
}

File: legacy/src/distributed/systems/grid/model/ResourceManager.java
package distributed.systems.grid.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;

import distributed.systems.grid.data.ActiveJob;
import distributed.systems.grid.simulation.SimulationContext;

/**
 * The entry point of a cluster
 *
 * Requirements:
 * 	Needs to hold the active jobs in some sort of way -> activeJobs
 *  Needs to know which Workers are available at all times -> through the hashmap
 *  Needs to know how to communicate with the workers -> through sockets
 *  
 * OnDie:
 * 	Whenever a resourceManager dies, no-one knows right away. I think that it is best that we detect this, in the schedulers. 
 * 	And do that by noticing that it does not send the confirmation message whenever we ask it to execute a job.
 *  Also this list of undead resourceManagers is scheduler dependent. 
 *    A.K.A every scheduler should privately keep track of which resourceManagers are alive.
 *    
 * OnResurrect:
 * 	If the resourceManager comes back to life, it might have an outdated / invalid workers list.
 *  but this will be fixed whenever the workers start sending their I am alive message again.
 */
public class ResourceManager extends GridNode {
	public static enum STATUS {
		AVAILABLE, RESERVED, BUSY, DEAD // TODO DEAD.
	}
	
	private STATUS status;
	private ArrayList<ActiveJob> activeJobs;

	/**
	 * This is an hashmap of the worker nodes
	 *  The idea is that since we cannot access the worker just by referencing their object 
	 *  We store the socket (which we use to communicate with the workers) and the status of that worker node
	 *    The resourceManager now knows at all times how to communicate with the workers, and what their status is.   
	 */
	private HashMap<Socket, Worker.STATUS> workers;

	public ResourceManager(SimulationContext context) {
		super(context, GridNode.TYPE.RESOURCE_MANAGER);

		this.status = STATUS.AVAILABLE;
		this.activeJobs = new ArrayList<ActiveJob>();
	}

	/**
	 * Triggered from the interface.
	 */
	public void toggleState() {
		this.status = this.status == STATUS.DEAD
			? STATUS.AVAILABLE
			: STATUS.DEAD;
	}

	/**
	 * The thread that checks if the workers are still alive
	 *  - WorkerOnDie -> send active job to other worker.
	 *  
	 *  TODO. check if workers are alive again
	 * 
	 * Note: this runs in an infinite loop, with 200ms sleep.
	 */
	public void runNode() {
		if (status == STATUS.DEAD) return;

		// for every active-unfinished job periodically check if the worker is still alive
		for (int i = 0; i < this.activeJobs.size(); i++) {
			if (this.activeJobs.get(i).getStatus() == Job.STATUS.RUNNING) {
				Socket worker = this.activeJobs.get(i).getWorker();
				
				// we did not hear from this guy for a loong time
				if (!worker.lastAliveIn(500L)) { // declare worker dead
					workers.put(worker, Worker.STATUS.DEAD);
					tryExecuteJob(this.activeJobs.get(i));
				}
				
				// did not hear from this worker in a slightly long time
				else if (!worker.lastAliveIn(200L)) { // request ping message
					sendPingRequestMessage(worker, this.activeJobs.get(i).getJob().getId());
				}
			}
		}
		
		// for every dead worker, check if he's alive again
		ArrayList<Socket> deadWorkers = getDeadWorkers();
		for (int i = 0; i < deadWorkers.size(); i++) {
			Socket worker = deadWorkers.get(i);
			if (!worker.lastAliveIn(10000L)) { // check if worker is alive only after long intervals
				worker.isAlive();
				sendRequestStatusMessage(worker);
			}
		}
	}

	/* ========================================================================
	 * 	Send messages below
	 * ===================================================================== */
	/**
	 * Check if the worker is still dead by sending it a status request message
	 */
	private void sendRequestStatusMessage(Socket worker) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.STATUS, 0, socket);
		worker.sendMessage(message);
	}
	
	/**
	 * confirmation message to the scheduler that it received the job correctly
	 */
	private void sendJobConfirmationToScheduler(Socket scheduler, int value) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.CONFIRMATION, value, socket);
		scheduler.sendMessage(message);
	}

	/**
	 * Request to a worker for it to start executing some code
	 */
	private void sendJobRequestToWorker(Socket worker, Job job) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.REQUEST, job.getId(), socket);
		message.attachJob(job);
		worker.sendMessage(message);
	}

	/**
	 * Confirmation to a worker that it received its result correctly
	 */
	private void sendJobResultConfirmationToWorker(Socket worker, int value) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.CONFIRMATION, value, socket);
		worker.sendMessage(message);
	}

	/**
	 * send job result back to the cluster
	 */
	private void sendJobResultToCluster(ActiveJob aj) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.RESULT, aj.getJob().getId(), socket);
		message.attachJob(aj.getJob());
		aj.getScheduler().sendMessage(message);
	}

	/**
	 * request job status from worker
	 */
	private void sendPingRequestMessage(Socket worker, int jobId) { 
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.PING, 0, socket);
		worker.sendMessage(message);
	}
	
	private void sendStatus(Socket recv) {
		if (this.status == STATUS.DEAD) return;

		Message message = new Message(Message.SENDER.RESOURCE_MANAGER, Message.TYPE.STATUS, 1, socket);
		recv.sendMessage(message);
	}

	/* ========================================================================
	 * 	Receive messages below
	 * ===================================================================== */
	/**
	 * Whenever the worker comes back to life (or updates it status for any other reason)
	 */
	private void workerStatusHandler(Message message) {
		Socket workerSocket = message.senderSocket; // we use the socket as identifier for the worker
		Worker.STATUS newStatus = Worker.STATUS.values()[message.getValue()]; // hacky way to convert int -> enum
		workers.put(workerSocket, newStatus);
		dequeueJobHandler();
	}

	/**
	 * whenever the RM receives a jobrequest:
	 * 	- adds the job to the activeJobs list
	 *  - it confirms the request to the scheduler
	 *  - try executing the job
	 */
	private void jobRequestHandler(Message message) {
		// System.out.println(">> ResourceManager has received new job -> starting to process job");
		ActiveJob aj = new ActiveJob(message.getJob(), message.senderSocket, null);
		this.activeJobs.add(aj);
		this.sendQueue(this.activeJobs.size());
		sendJobConfirmationToScheduler(message.senderSocket, message.getValue());

		tryExecuteJob(aj);
	}

	/**
	 * When the job gets confirmed by the worker (meaning that he is going to work on it)
	 *  - update the jobstatus to RUNNING
	 *  - update lastseen of worker
	 *  - update worker to BUSY
	 */
	private void jobConfirmationHandler(Message message) {
		workers.put(message.senderSocket, Worker.STATUS.BUSY); // removes the reserved status
		ActiveJob aj = getActiveJob(message.getValue());
		aj.setStatus(Job.STATUS.RUNNING);
		message.senderSocket.isAlive(); // update alive time of worker
	}

	/**
	 * When we receive a jobresult from a worker node
	 */
	public void jobResultHandler(Message message) {
		// System.out.println("<< ResourceManager done with job");
		int jobId = message.getValue();
		ActiveJob aj = getActiveJob(jobId);
		aj.setStatus(Job.STATUS.CLOSED);
		aj.setJob(message.getJob());

		workers.put(message.senderSocket, Worker.STATUS.AVAILABLE); // set worker status to available for RM 
		sendJobResultConfirmationToWorker(message.senderSocket, jobId); // worker will update its internal status after receiving this message
		
		sendJobResultToCluster(aj);
	}

	/**
	 * When the scheduler confirms the result 
	 */
	public void jobResultConfirmationHandler(Message message) {
		int jobId = message.getValue();
		ActiveJob aj = getActiveJob(jobId);
		this.activeJobs.remove(aj);
		this.sendQueue(this.activeJobs.size());
	}
	
	/**
	 * when we receive the 'hey im still alive message from a worker'
	 *  - update the last seen status of worker
	 *  
	 *  - actually not necessary since it happens in the onMessageReceived func.
	 */
	public void jobPingResultHandler(Message message) {
		Socket worker = message.senderSocket;
		worker.isAlive();		
	}
	
	public void pingRequestHandler(Message message) {
		Socket scheduler = message.senderSocket;
		sendStatus(scheduler);
	}

	/**
	 * Types of messages we expect here:
	 * 
	 * - Message from a worker saying that he is available
	 * - Message from a worker saying that he received the Job and is going to start working on it. 
	 * - Message from a worker saying that he is done calculating and gives you back the results.
	 * - Message from a scheduler saying that he has a Job for the cluster
	 * - Message from a scheduler saying that he received the result of the Job correctly 
	 */
	public void onMessageReceived(Message message) {
		
		message.senderSocket.isAlive(); // notice that sender is still alive
		if (message.getSender() == Message.SENDER.WORKER) {
			if (message.getType() == Message.TYPE.STATUS) {
				workerStatusHandler(message);
				return;
			}
			if (message.getType() == Message.TYPE.CONFIRMATION) {
				jobConfirmationHandler(message);
				return;
			}
			if (message.getType() == Message.TYPE.RESULT) {
				jobResultHandler(message);
				return;
			}
			if (message.getType() == Message.TYPE.PING) {
				jobPingResultHandler(message);
				return;
			}
		}

		if (message.getSender() == Message.SENDER.SCHEDULER) {
			if (message.getType() == Message.TYPE.REQUEST) {
				jobRequestHandler(message);
				return;
			}
			if (message.getType() == Message.TYPE.ACKNOWLEDGEMENT) {
				jobResultConfirmationHandler(message);
				return;
			}
			if (message.getType() == Message.TYPE.PING) {
				pingRequestHandler(message);
				return;
			}
		} 

		// exception
	}

	/* ========================================================================
	 * 	Class functions
	 * ===================================================================== */
	/**
	 * Whenever a worker becomes available, we should check if there is a job in queue that can be executed.
	 */
	private void dequeueJobHandler() {
		ActiveJob aj = getQueuedJob();
		if (aj == null) {
			return;
		}

		tryExecuteJob(aj);
	}

	private void tryExecuteJob(ActiveJob aj) {
		if (status == STATUS.DEAD) {
			// dead man cant jump
		}
		
		Socket availableWorker = getAvailableWorker();

		if (availableWorker == null) {
			// status remains waiting -> whenever a node worker becomes available, give him this job.
			return;
		}

		aj.setStatus(Job.STATUS.RUNNING);
		workers.put(availableWorker, Worker.STATUS.RESERVED);
		availableWorker.isAlive(); // say that this second was the last time we say the worker alive (so that it does not pollute the network too fast)
		aj.setWorker(availableWorker);
		sendJobRequestToWorker(availableWorker, aj.getJob());
	}

	private Socket getAvailableWorker() {
		for (Entry<Socket, Worker.STATUS> entry : workers.entrySet()) {
			Worker.STATUS status = entry.getValue();
			if (status == Worker.STATUS.AVAILABLE) {
				return entry.getKey();
			}
		}

		return null;
	}

	public ActiveJob getActiveJob(int jobId) {
		for (int i = 0; i < this.activeJobs.size(); i++) {
			if (this.activeJobs.get(i).getJob().getId() == jobId) {
				return this.activeJobs.get(i);
			}
		}
		return null;
	}

	public ActiveJob getQueuedJob() {
		for (int i = 0; i < this.activeJobs.size(); i++) {
			if (this.activeJobs.get(i).getStatus() == Job.STATUS.WAITING) {
				return this.activeJobs.get(i);
			}
		}
		return null;
	}

	public void setWorkerSockets(List<Socket> workers) {
		this.workers = new HashMap<Socket, Worker.STATUS>();
		for (int i = 0; i < workers.size(); i ++) {
			this.workers.put(workers.get(i), Worker.STATUS.AVAILABLE);
		}
	}
	
	public ArrayList<Socket> getDeadWorkers(){
		ArrayList<Socket> deadWorkers = new ArrayList<Socket>();
		
		for (Entry<Socket, Worker.STATUS> entry : workers.entrySet()) {
			Worker.STATUS status = entry.getValue();
			if (status == Worker.STATUS.DEAD) {
				deadWorkers.add(entry.getKey());
			}
		}
		
		return deadWorkers;
	}
}
